openhands/memory/condenser.py

```diff
@@ -1,24 +1,412 @@
+
+        self._condensation = results
+        self._last_history_length = len(state.history)
+
+        return results
+
+
+@dataclass
+class CondensationObservation(Observation):
+    """The output of a condensation action."""
+
+    observation: str = ObservationType.RUN
+
+    @property
+    def message(self) -> str:
+        return self.content
+
+
+class NoOpCondenser(Condenser):
+    """A condenser that does nothing to the event sequence."""
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Returns the list of events unchanged."""
+        return events
+
+
+class ObservationMaskingCondenser(Condenser):
+    """A condenser that masks the values of observations outside of a recent attention window."""
+
+    def __init__(self, attention_window: int = 5):
+        self.attention_window = attention_window
+
+        super().__init__()
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Replace the content of observations outside of the attention window with a placeholder."""
+        results: list[Event] = []
+        for i, event in enumerate(events):
+            if (
+                isinstance(event, Observation)
+                and i < len(events) - self.attention_window
+            ):
+                results.append(CondensationObservation('<MASKED>'))
+            else:
+                results.append(event)
+
+        return results
+
+
+class RecentEventsCondenser(Condenser):
+    """A condenser that only keeps a certain number of the most recent events."""
+
+    def __init__(self, keep_first: int = 0, max_events: int = 10):
+        self.keep_first = keep_first
+        self.max_events = max_events
+
+        super().__init__()
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Keep only the most recent events (up to `max_events`)."""
+        head = events[: self.keep_first]
+        tail_length = max(0, self.max_events - len(head))
+        tail = events[-tail_length:]
+        return head + tail
+
+
+class LLMSummarizingCondenser(Condenser):
+    """A condenser that relies on a language model to summarize the event sequence as a single event."""
+
+    def __init__(self, llm: LLM):
+        self.llm = llm
+
+        super().__init__()
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Applies an LLM to summarize the list of events.
+
+        Raises:
+            Exception: If the LLM is unable to summarize the event sequence.
         """
         try:
-            messages = [{'content': summarize_prompt, 'role': 'user'}]
-            resp = llm.completion(messages=messages)
-            summary_response = resp['choices'][0]['message']['content']
-            return summary_response
+            # Convert events to a format suitable for summarization
+            events_text = '\n'.join(f'{e.timestamp}: {e.message}' for e in events)
+            summarize_prompt = f'Please summarize these events:\n{events_text}'
```

Comment:
Just curious, not a review comment, did you also test this strategy?

Reply:
I did, and it didn't resolve a single SWE-bench instance on my test subset. The agent kept getting stuck in loops. My guess is it needs some serious prompt-tuning.

Reply:
Ah, I see, thanks!

Yes, I'm personally convinced that the prompts were a major reason why the results weren't great in the past. I will absolutely love if we can do something like, set up a prompt eval framework, or site, or whatever form. If you want any kind of help on such thing, feel free to just let me know.